scalyr_agent/builtin_monitors/windows_event_log_monitor.py

```diff
@@ -773,6 +801,91 @@ def log_event(self, event):
         finally:
             self._bookmark_lock.release()
 
+    def _param_placeholder_value(self, channel, provider, param):
+        if not re.match("^%%[0-9]+$", param):
+            return param
+
+        if channel == "Security":
+            provider = "Security"
+
+        param_key = "%s-%s-%s" % (channel, provider, param)
+        if param_key in self._param_cache:
+            return self._param_cache[param_key]
+
+        dll_key = "%s-%s" % (channel, provider)
+        if dll_key in self._dll_cache:
+            if not isinstance(self._dll_cache[dll_key], _DLL):
+                return param
+        else:
+            # Before opening a new handle to a dll check if another handle to the same dll is open;
+            # multiple event log providers may ultimately reference the same dll.
+
+            try:
+                dllpath = _DLL.dllpath(channel, provider)
+            except Exception as e:
+                self._dll_cache[dll_key] = e
+                return param
+
+            found = False
+            for dll in self._dll_cache.values():
+                if dll.path == dllpath:
+                    self._dll_cache[dll_key] = dll
+                    found = True
+                    break
+
+            if not found:
+                try:
+                    self._dll_cache[dll_key] = _DLL(channel, provider)
+                except Exception as e:
+                    self._dll_cache[dll_key] = e
+                    return param
+
+        try:
+            value = win32api.FormatMessageW(
+                win32con.FORMAT_MESSAGE_FROM_HMODULE,
+                self._dll_cache[dll_key].handle,
+                int(param[2:]),
+                self._langid,
+                None,
+            )
+            self._param_cache[param_key] = value.strip()
+        except:
```

Comment:
Probably a good idea to log a message here with the actual exception + context? At very least, under debug in case it doesn't warrant info / warn log level.

Reply:
Will do